File: src/my_py_pkg/my_py_pkg/add_two_ints_client_no_oop.py
#!/usr/bin/env python3
# Shebang line specifying the interpreter to be used for executing the script

import rclpy
# Importing the rclpy module for working with ROS 2 Python

from rclpy.node import Node
# Importing the Node class from rclpy.node module

from example_interfaces.srv import AddTwoInts
# Importing the AddTwoInts service message from the example_interfaces.srv module

def main(args=None):
    # Entry point of the script

    rclpy.init(args=args)
    # Initializing the ROS 2 Python client library

    node = Node("add_two_ints_no_opp")
    # Creating a Node object with the name "add_two_ints_no_opp"

    client = node.create_client(AddTwoInts, "add_two_ints")
    # Creating a client for the AddTwoInts service with the service name "add_two_ints"

    while not client.wait_for_service(1.0):
        # Waiting for the service to become available
        node.get_logger().Warning("Waiting for server Add Two Ints")

    request = AddTwoInts.Request()
    # Creating a request object of type AddTwoInts.Request

    request.a = 3
    request.b = 8
    # Setting the values of 'a' and 'b' in the request object

    future = client.call_async(request)
    # Calling the service asynchronously and getting a future object

    rclpy.spin_until_future_complete(node, future)
    # Spinning the node until the future is complete

    try:
        response = future.result()
        # Getting the response from the completed future

        node.get_logger().info(str(request.a) + " + " + str(request.b) + " = " + str(response.sum))
        # Logging the sum of 'a' and 'b' received in the response

    except Exception as e:
        node.get_logger().error("Service call failed %r" % (e,))
        # Logging an error message if the service call fails

    rclpy.shutdown()
    # Shutting down the ROS 2 Python client library

if __name__ == "__main__":
    main()
    # Calling the main function if the script is run directly

File: src/my_cpp_pkg/src/add_two_ints_client_no_oop.cpp
#include "rclcpp/rclcpp.hpp"                       // Include the RCLCPP library
#include "example_interfaces/srv/add_two_ints.hpp" // Include the AddTwoInts service message

int main(int argc, char **argv)
{
    rclcpp::init(argc, argv);
    auto node = std::make_shared<rclcpp::Node>("add_two_ints_client");
    auto client = node->create_client<example_interfaces::srv::AddTwoInts>("add_two_ints"); // Create a client for the AddTwoInts service
    while (!client->wait_for_service(std::chrono::seconds(1)))                              // Wait for the service to be available
    {
        if (!rclcpp::ok()) // Check if ROS 2 is still running
        {
            RCLCPP_ERROR(node->get_logger(), "Interrupted while waiting for the service. Exiting.");
            return 0; // Return without continuing execution
        }
        RCLCPP_INFO(node->get_logger(), "Service not available, waiting again...");
    }
    for (int i = 1; i <= 5; i++)
    {
        auto request = std::make_shared<example_interfaces::srv::AddTwoInts::Request>(); // Create a request object
    request->a = 3+i;                                                                  // Set the request values
    request->b = 4-2*i;

    auto future = client->async_send_request(request); // Send the request asynchronously

    if (rclcpp::spin_until_future_complete(node, future) == rclcpp::FutureReturnCode::SUCCESS) // Wait for the response
    {
        auto response = future.get();                               // Get the response
        RCLCPP_INFO(node->get_logger(), "Sum: %ld", response->sum); // Print the result
    }
    else            //if the request fails
    {
        RCLCPP_ERROR(node->get_logger(), "Failed to call service");
    }


    }
    
    rclcpp::shutdown();
    return 0;
}
